Add BackgroundSizing

robloo opened the pull request:
## What does the pull request do?

 1. Adds a new `BackgroundSizing` enum based on UWP/WinUI
     - Aside from the WinUI-defined `InnerBorderEdge` and `OuterBorderEdge` values, the enum also adds a new `CenterBorder` value (the default) to support backwards-compatibility with the way Avalonia worked before this PR.
 1. The following controls were updated to add the `BackgroundSizing` property.
     - Border
     - TemplatedControl
     - ContentControl (automatically binds the property with the default template)
     - ContentPresenter
 1. A new `GeometryBuilder` helper class was created (based on the one in WinUI). Related methods were moved here from a few different places namely `BorderRenderHelper` and the `Rectangle` shape.
 1. `BorderRenderHelper` was updated to use the `GeometryBuilder` and support `BackgroundSizing`. Related control's rendering code was changed to use it.
 1. As a bonus with `GeometryBuilder`, `RectangleGeometry` has been updated to support `RadiusX`/`RadiusY` and the `Rectangle` shape now uses it directly (more functionality plus shared code).

## What is the current behavior?

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/f1ddc13e-1c45-427d-81be-67d2f47b137e)

 * Note the rendering errors (a bug)
 * Unfortunately, there is a combination of `CenterBorder` (simple path) and `InnerBorderEdge` (complex path) based on the rendering path taken. This is a bug as well as rendering shouldn't differ like this.

## What is the updated/expected behavior with this PR?

<details>
<summary>XAML (Modified BorderPage.xaml)</summary>

```XAML
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             d:DesignHeight="800"
             d:DesignWidth="400"
             x:Class="ControlCatalog.Pages.BorderPage">
  <StackPanel Orientation="Vertical" Spacing="4">
    <TextBlock Classes="h2">A control which decorates a child with a border and background</TextBlock>

    <StackPanel Orientation="Vertical"
                Margin="0,16,0,0"
                HorizontalAlignment="Center"
                Spacing="16">
      <Border BorderBrush="{DynamicResource SystemAccentColor}" BorderThickness="2" Padding="16">
        <TextBlock>Border</TextBlock>
      </Border>
      <Border Background="{DynamicResource SystemAccentColorDark1}" 
              BorderBrush="{DynamicResource SystemAccentColor}" 
              BorderThickness="4" 
              Padding="16">
        <TextBlock>Border and Background</TextBlock>
      </Border>
      <Border BorderBrush="{DynamicResource SystemAccentColor}"
              BorderThickness="4" 
              CornerRadius="8"
              Padding="16">
        <TextBlock>Rounded Corners</TextBlock>
      </Border>
      <Border Background="{DynamicResource SystemAccentColor}"
              CornerRadius="8"
              Padding="16">
        <TextBlock>Rounded Corners</TextBlock>
      </Border>
      <Border Width="100" Height="100"
              BorderThickness="0"
              Background="White"
              CornerRadius="100" ClipToBounds="True">
        <Image Source="/Assets/maple-leaf-888807_640.jpg" Stretch="UniformToFill" />
      </Border>
      <TextBlock Text="Border with Clipping" HorizontalAlignment="Center" />
    </StackPanel>

    <StackPanel Orientation="Horizontal"
                Spacing="16">
      <Border Width="200"
              Background="Blue"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#5000FF00"
              BorderThickness="20">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Blue"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#5000FF00"
              BorderThickness="20">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Blue"
              BackgroundSizing="CenterBorder"
              BorderBrush="#5000FF00"
              BorderThickness="20">
        <TextBlock Text="CenterBorder" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
                Spacing="16">
      <Border Width="200"
              Height="100"
              Background="Blue"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#5000FF00"
              BorderThickness="15"
              CornerRadius="30">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Height="100"
              Background="Blue"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#5000FF00"
              BorderThickness="15"
              CornerRadius="30">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Height="100"
              Background="Blue"
              BackgroundSizing="CenterBorder"
              BorderBrush="#5000FF00"
              BorderThickness="15"
              CornerRadius="30">
        <TextBlock Text="CenterBorder" />
      </Border>
    </StackPanel>

    <!-- InnerBorderEdge -->
    <StackPanel Orientation="Horizontal"
                Spacing="16"
                Margin="0,30,0,0">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0"
              CornerRadius="20">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0"
              CornerRadius="20">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0"
              CornerRadius="20">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20"
              CornerRadius="20">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="20,0,0,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,20,0,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,0,20,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,0,0,20">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0"
              CornerRadius="20,0,0,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0"
              CornerRadius="0,20,0,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0"
              CornerRadius="0,0,20,0">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="InnerBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20"
              CornerRadius="0,0,0,20">
        <TextBlock Text="InnerBorderEdge" />
      </Border>
    </StackPanel>

    <!-- OuterBorderEdge -->
    <StackPanel Orientation="Horizontal"
                Spacing="16"
                Margin="0,30,0,0">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0"
              CornerRadius="20">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0"
              CornerRadius="20">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0"
              CornerRadius="20">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20"
              CornerRadius="20">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="20,0,0,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,20,0,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,0,20,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,0,0,20">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0"
              CornerRadius="20,0,0,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0"
              CornerRadius="0,20,0,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0"
              CornerRadius="0,0,20,0">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="OuterBorderEdge"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20"
              CornerRadius="0,0,0,20">
        <TextBlock Text="OuterBorderEdge" />
      </Border>
    </StackPanel>

    <!-- CenterBorder -->
    <StackPanel Orientation="Horizontal"
                Spacing="16"
                Margin="0,30,0,0">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20">
        <TextBlock Text="CenterBorder" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0"
              CornerRadius="20">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0"
              CornerRadius="20">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0"
              CornerRadius="20">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20"
              CornerRadius="20">
        <TextBlock Text="CenterBorder" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="20,0,0,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,20,0,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,0,20,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="20"
              CornerRadius="0,0,0,20">
        <TextBlock Text="CenterBorder" />
      </Border>
    </StackPanel>

    <StackPanel Orientation="Horizontal"
            Spacing="16">
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="20,0,0,0"
              CornerRadius="20,0,0,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,20,0,0"
              CornerRadius="0,20,0,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,20,0"
              CornerRadius="0,0,20,0">
        <TextBlock Text="CenterBorder" />
      </Border>
      <Border Width="200"
              Background="Green"
              BackgroundSizing="CenterBorder"
              BorderBrush="#7FFF0000"
              BorderThickness="0,0,0,20"
              CornerRadius="0,0,0,20">
        <TextBlock Text="CenterBorder" />
      </Border>
    </StackPanel>
  </StackPanel>
</UserControl>

```

</details>

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/44eef0bf-d062-46c9-871c-aa7dfcd8a421)

## How was the solution implemented (if it's not obvious)?

First, an attempt was made with a completely-custom algorithm not based on WPF or WinUI. This has a few advantages listed below but was decided against due to the rendering differences (which I still think Microsoft made an error with).

 1. Calculating a corner radius for an entire side of the rounded rectangle produces strange artifacts in WinUI. I mean these little hooks that seem to protrude past the actual edge you have a border thickness set on. This looks like a bug to me and it doesn't exist in the custom algorithm -- how can a corner radius on a border corner cause the border to extend past it's defined edge thickness?
 1. Calculating the inside corner radius is different. WinUI has an algorithm that is calculating an insider radius even if there isn't enough space for it using a naive approach. With the custom current algorithm the inner corner radius is just crushed to zero and the insider becomes a simple rectangle in this case. Note this can be fixed but the code was abandoned due to num.1 above.

<details>
<summary>Custom Algorithm Screenshot</summary>

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/a07c9d50-cf17-4b00-b932-41ad8e91cd29)

</details>

Finally, the WinUI source code was followed for the key algorithm to reproduce the functionality as much as possible. ~Note this could not be followed exactly since WinUI uses a D2D boolean geometry calculation we don't have in Avalonia.~

 * [geometry.cpp CGeometryBuilder::CalculateRoundedCornersRectangle](https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/geometry.cpp#L862-L869)
 * [framework.cpp CFrameworkElement::CreateBorderGeometriesAndBrushClipsCommon](https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/framework.cpp#L3693-L3701)

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes

None*

*Differences
 - The default `BackgroundSizing` is set to a new `CenterBorder` value which should very closely reproduce the existing behavior in most cases. However, the existing behavior in Avalonia has bugs and uses CenterBorder for the simple rendering path and InsideBorderEdge for the complex path. So to fix this error there is no way but to make one of them different. In this case CenterBorder and the simple rendering path is far more widely used so is the default (and original Avalonia functionality). The simple rendering path will not have any differences.
 - At extremes in the complex render path it's possible for the geometries to be slightly different.
   - There are exising bugs in Avalonia (see above) that need to be fixed
   - ~The paths (especially combined border path) is calculated slightly different than upstream WinUI. This cannot be changed without adding support for boolean path geometry calculations (intersection) either within Avalonia itself or as a new dependency. See the [WinUI 3 source code](https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/framework.cpp#L3799-L3822) for more information:~ Edit: Calculation with CombinedGeometry was in the end the same.

## Obsoletions / Deprecations

None

## Fixed issues

Closes #9946
Closes #13725
Fixes #11490

robloo committed:
Add BackgroundSizing enum

robloo commented on src/Avalonia.Base/Media/BackgroundSizing.cs:
```diff
@@ -0,0 +1,38 @@
+    /// <summary>
+    /// Defines how a background is drawn relative to its border.
+    /// </summary>
+    public enum BackgroundSizing
```
I debated if this should be in Avalonia.Media (base) or Avalonia.Controls. Finally it was decided to go in the lowest level that made sense.

 - It is useful in the new GeometryBuilder which should go with the other Geometry types in Avalonia.Media
 - WinUI upstream has this defined in DirectUI even lower than the UI framework: https://github.com/microsoft/microsoft-ui-xaml/blob/93742a178db8f625ba9299f62c21f656e0b195ad/dxaml/xcp/core/core/elements/framework.cpp#L3719C39-L3719C47

robloo committed:
Improve formatting in RoundedRect

robloo committed:
Add the BackgroundSizing property to relevant controls

robloo committed:
Add new GeometryBuilder helper

This is based on new ideas and APIs in both WinUI and WPF

robloo committed:
Support BackgroundSizing in BorderRenderHelper

This also removes the last remnants of dash array support in the internal APIs

robloo committed:
Remove old code in BorderRenderHelper moved to GeometryBuilder

robloo committed:
Move Rectangle shape geometry calculation into GeometryBuilder

robloo committed:
Add RadiusX/Y properties to RectangleGeometry

robloo committed:
Use RectangleGeometry directly in the Rectangle shape

Since RectangleGeometry now supports RadiusX/Y there is no longer a need to have a separate geometry creation path.

robloo commented on src/Avalonia.Base/Media/RectangleGeometry.cs:
```diff
@@ -8,6 +7,18 @@ namespace Avalonia.Media
     /// </summary>
     public class RectangleGeometry : Geometry
     {
+        /// <summary>
+        /// Defines the <see cref="RadiusX"/> property.
+        /// </summary>
+        public static readonly StyledProperty<double> RadiusXProperty =
+            AvaloniaProperty.Register<RectangleGeometry, double>(nameof(RadiusX));
+
+        /// <summary>
+        /// Defines the <see cref="RadiusY"/> property.
+        /// </summary>
+        public static readonly StyledProperty<double> RadiusYProperty =
+            AvaloniaProperty.Register<RectangleGeometry, double>(nameof(RadiusY));
```
Note `RectangleGeometry`, `Rectangle` (and even `EllipseGeometry` each defined `RadiusX` and `RadiusY` properties independently. There is no usage of AddOwner. This follows what is done in WPF and makes sense I think due to the layering here between Controls and more core/base types like geometry.

maxkatz6 replied:
I understand separating Ellipse*** and Rectangle*** properties completely, but not sure about Rectangle and RectangleGeometry? 

On other hand, I don't see any use cases when it's realistically useful to reuse the same property here.

robloo replied:
A geometry (RectangleGeometry) is conceptually quite a bitter different than a control/shape (Rectangle). Their use cases are different and implementations are very different. They exist at different levels of the framework for different things. So to me it makes sense not to use AddOwner -- but I expected some push back so made the note.

Also, when in doubt, I follow upstream:

 - [WPF `RectangleGeometry`](https://github.com/dotnet/wpf/blob/c79cdd3ef07fc3cb8954178e5c5decfb51f237ca/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/Generated/RectangleGeometry.cs#L380-L397) Note that I'm not sure why these properties are auto-generated. Perhaps that influenced the original decision.
 - [WPF `Rectangle` shape](https://github.com/dotnet/wpf/blob/c79cdd3ef07fc3cb8954178e5c5decfb51f237ca/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Shapes/Rectangle.cs#L51-L101)

robloo committed:
Add WinUI-based CalculateRoundedCornersRectangle() method

robloo committed:
Simplify and use the WinUI-algorithm in BorderRenderHelper

robloo committed:
Update GeometryBuilder.DrawRoundedCornersRectangle() based on WinUI

robloo committed:
Remove obsolete WPF-based CalculateRoundedCornersRectangle method

robloo committed:
Update CalculateRoundedCornersRectangleAlternate

robloo committed:
Update GeometryBuilderTests

maxkatz6 commented:
Note: if anything about corner radius algorithm is changed, this code should also be double checked: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/BorderVisual.cs#L48

robloo commented:
> Note: if anything about corner radius algorithm is changed, this code should also be double checked: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/BorderVisual.cs#L48

@maxkatz6 Yea, I saw that. It only affects the simple path and is easy to add back those changes.

I had to create a branch several commits back for now because a commit was made that requires the latest version of Visual Studio. I don't want to go through and update my entire dev setup quite yet.

maxkatz6 commented:
> because a commit was made that requires the latest version of Visual Studio

What's the problem specifically?

robloo commented:
> > because a commit was made that requires the latest version of Visual Studio
> 
> What's the problem specifically?

@maxkatz6 The commits in PR#13719 changed global.json to require `"version": "7.0.404"` on dev machines. Not only does that require the new SDK but .NET 7.0.404 requires Visual Studio version 17.7+ I think (previously it was 17.4+). That requires updates to my dev setup I just haven't done and didn't want to do at the moment. It isn't a showstopper I just branched before that commit to avoid long Visual Studio and SDK updates that day. Sometimes I just don't want to play around for an hour getting the build to work again.

robloo committed:
Merge branch 'master' into robloo/background-sizing

robloo committed:
Fix merge

robloo committed:
Remove unused dashed line properties in BorderRenderHelper

These were already removed in master but could not be merged directly because of the new backgroundSizing parameter.

robloo committed:
Optimize CalculateRoundedCornersRectangle

In the worse-case up to three Point structs were created during calculation for each keypoint. Now doubles are used during calculations so the Point is created once at the end.

robloo commented:
@maxkatz6 

> Note: if anything about corner radius algorithm is changed, this code should also be double checked:  https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/BorderVisual.cs#L48

The way the corner radius is calculated (i.e. what it means) shouldn't be changing. It's the radius at the center of the stroke/border thickness. That said, it does seem plausible that the referenced code in BorderVisual would need to be aware of BackgroundSizing to adjust the RoundedRect size accordingly. I suspect there might be some pre-existing bugs here too in the way things are handled. Someone familiar with this area of the code really should look it over.

This also starts getting into the topic of #2105 which can also now be solved with the GeometryBuilder generated geometries. Since RectangleGeometry now supports simple corner-radii it could be used for clipping in simple scenarios as well. However, since Microsoft never standardized between elliptical, circular, uniform/non-uniform corner-radii (https://github.com/AvaloniaUI/Avalonia/discussions/7923) there are cases RectangleGeometry can't handle and the full path geometry is needed.

robloo committed:
Adjust corner radius calculation within custom GeometryBuilder.CalculateRoundedCornersRectangle()

Corner radius is defined at the center of the border stroke rather than the outside edge.

robloo committed:
Simplify code in WinUI's algorithm for CalculateRoundedCornersRectangle()

robloo committed:
Remove extra BorderRenderHelper.RenderCore method

This method does not need to be separate from Render.

robloo committed:
Add more CalculateRoundedCornersRectangle tests and adjust method naming

robloo committed:
Remove custom CalculateRoundedCornersRectangle() algorithm and always use the WinUI one

robloo committed:
Merge branch 'master' into robloo/background-sizing

maxkatz6 commented on src/Avalonia.Base/Media/GeometryBuilder.cs:
```diff
@@ -0,0 +1,610 @@
+        /// </remarks>
+        /// <param name="context">The geometry context to draw into.</param>
+        /// <param name="keypoints">The rounded rectangle keypoints defining the rectangle to draw.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void DrawRoundedCornersRectangle(
```
I wonder why this method has AggressiveInlining. It probably won't be much efficient, since these methods are rather big.

I am going to try to run this PR and review in detail on the next week.

robloo replied:
Yea, I'm not sure why that's there either at this point. I'll remove it.

robloo commented on src/Avalonia.Base/Media/GeometryBuilder.cs:
```diff
@@ -0,0 +1,610 @@
+        /// These keypoints can be combined together to produce geometries for both background
+        /// and border elements.
+        /// </remarks>
+        internal struct RoundedRectKeypoints
```
Note that this was switched from a class to a struct for theoretical performance reasons. However, it is now mutable.

The original class was in BorderRenderHelper called `BorderGeometryKeypoints`.

maxkatz6 replied:
Hm, it's an interesting one. Changing it to struct would play well, if this struct was only used in a scope of the single method. But since it's a return type of the method, and also an argument of another, it might get copied way too much.

robloo replied:
Yea, point taken. There are a lot of members in `RoundedRectKeypoints`. I was trying to avoid the overhead of a class here though.

Since this is an internal API I see no problem in passing by `ref` to the related methods to avoid the copy. Then we get both optimizations. Seems appropriate considering this is fairly low level and graphics-related.

robloo committed:
Remove some AggressiveInlining attributes

robloo committed:
Make parameters clear

robloo committed:
Pass RoundedRectKeypoints by ref

robloo commented on src/Avalonia.Controls/Utils/BorderRenderHelper.cs:
```diff
@@ -63,17 +70,25 @@ void Update(Size finalSize, Thickness borderThickness, CornerRadius cornerRadius
                 {
-                    var borderGeometryKeypoints =
-                        new BorderGeometryKeypoints(boundRect, borderThickness, cornerRadius, false);
                     var borderGeometry = new StreamGeometry();
```
@maxkatz6 I just noticed with #14360 we do actually have geometry Boolean operations implemented. The heavy lifting is done with Skia and it was added in #6576.

I need to use this to calculate the border geometry below. That is what WinUI does and it should make the algorithm exactly the same here. Below (even with the old code) we are relying on an undefined implementation detail to combine the two geometries in the way we want. It's going to be much cleaner not to do this and it should also remove the last little artifacts visible in the exaggerated cases.

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/de24ce69-9be0-461d-85e6-687fbdb0f8de)

robloo replied:
Updated and it looks better now

robloo committed:
Use GeometryCombineMode.Exclude to better calculate border geometries

maxkatz6 committed:
Merge branch 'master' into robloo/background-sizing

maxkatz6 committed:
Update BorderPage.xaml to include BackgroundSizing API

maxkatz6 reviewed (approved):
LGTM

I was a bit worried about new background sizing options, as they might have measurably worse performance, if enabled by default. But after quick run with dotTrace, it seems I was worried for not reason.
Some proper benchmarks might be useful to have, and we already have some basic infrastructure for that. But we don't need it in this PR.

Also pushed an updated BorderPage with couple of new borders just to have something. We probably don't need full page from your description :)

maxkatz6 committed:
Merge branch 'master' into robloo/background-sizing

robloo commented:
> Also pushed an updated BorderPage with couple of new borders just to have something. We probably don't need full page from your description :)

That's fine. I included the XAML in the PR description for testing in the future but as you said it was too much to add so I didn't. I'm fine with just a few examples added.

maxkatz6 committed:
Merge branch 'master' into robloo/background-sizing

robloo commented:
@maxkatz6 

> I was a bit worried about new background sizing options, as they might have measurably worse performance, if enabled by default. But after quick run with dotTrace, it seems I was worried for not reason.

I forgot to mention: I think a lot of this is due to the caching of the geometries. They don't need to be rebuilt very often so there is little performance impact. It doesn't need to be recalculated during a normal refresh.

That said, the code works almost exactly the same as it used to, there is little additional complexity to support these modes in the complex render path except one thing: creating and combining the separate border geometries at the end using Exclude. However, geometry combining is done almost directly by Skia and is quite fast. StreamGeometry creation is pretty fast too and we use it for a lot of low level stuff already (this is only done one additional time now).

maxkatz6 commented:
> That said, the code works almost exactly the same as it used to, there is little additional complexity to support these modes

Yeah, I know that. At least for default CenterBorder value. Two new values do add a little bit overhead, but as you mentioned - it's all cached and fast.


A while ago we were thinking about possibility to render Borders directly in the Skia using precompiled shaders. Where we would completely skip whole geometry step and only pass basic parameters to the shader, and it will draw everything on pixel shading step. But this solution would be less portable and quite involving, so was forgotten. I think new Flutter rendering engine does it similarly as well.
